"borrow checker invariants" section of the (#2867)
google/comprehensive-rust: src/idiomatic/leveraging-the-type-system/borrow-checker-invariants/single-use-values.md

---
minutes: 0
---

# Single-use values

In some circumstances we want values that can be used _exactly once_. One critical example of this is in cryptography: "Nonces."

```rust,editable
fn main() {
pub struct Key;

// Pretend this is a cryptographically unique, use-once number.
pub struct Nonce(u32);

// It's unsafe to declare a nonce directly! In practice,
// this would be done with an RNG source, and potentially
// a timestamp.
unsafe fn new_nonce_from_raw(nonce: u32) -> Nonce {
    Nonce(nonce)
}

let nonce = unsafe { new_nonce_from_raw(1337) };
let data_1: [u8; 4] = [1, 2, 3, 4];
let data_2: [u8; 4] = [4, 3, 2, 1];
let key = Key;

// The key and data can be re-used, copied, etc. but the nonce cannot.
fn encrypt(nonce: Nonce, key: &Key, data: &[u8]) {}

encrypt(nonce, &key, &data_1);
encrypt(nonce, &key, &data_2); // 🛠️❌
}

> ```

Reviewer: I think the `new_non_from_raw`/`unsafe` aspect add unnecessary noise to the example. We don't need to go into the details as to how a `Nonce` is constructed, we just need a constructor/generator to exist.

<details>

- Owned "consumption" lets us model single-once values.

- Not implementing clone/copy here and making the interior type opaque (as per the newtype pattern) is _intentional_, as it prevents multiple uses of the same, API-controlled value.

- I.e. A Nonce is a additional piece of random, unique data during an encryption process that helps prevent "replay attacks".

- In practice people have ended up re-using nonces in circumstances where security is important, making it possible for private key information to be derived by attackers.

- By tying nonce creation and consumption up in rust's ownership model, and by not implementing clone/copy on sensitive single-use data, we can prevent this kind of dangerous misuse.

> - Cryptography Nuance: There is still the case where a nonce may be used twice if it's created through purely a pseudo-random process with no additional metadata, and that circumstance can't be avoided through this particular method. This kind of API prevents one kind of misuse, but not all kinds.

Author: That was what I was aiming to do, but I got a bit lost in the weeds 😅

    let data_2: [u8; 4] = [4, 3, 2, 1];
    let key = Key(/* specifics omitted */);

    // The key and data can be re-used, copied, etc. but the nonce cannot.
    encrypt(nonce, &key, &data_1);
    // encrypt(nonce, &key, &data_2); // 🛠️❌
}
```

<details>

- Owned "consumption" of values lets us model things that need to be single-use.

- By keeping constructors private and not implementing clone/copy for a type,
  making the interior type opaque (as per the newtype pattern), we can prevent
  multiple uses of the same value while keeping control of how that value is
  bui.

- A nonce is a piece of random, unique data used in cryptographic protocols to
  prevent replay attacks.

  - In practice people have ended up accidentally re-using nonces. Most
    commonly, this causes the cryptographic protocol to completely break down
    and stop fulfilling its function. Depending on the specifics of nonce reuse
    and cryptography at hand, private keys can also become computable by
    attackers.

  - By tying nonce creation and consumption up in rust's ownership model, and by
    not implementing clone/copy on sensitive single-use data, we can prevent

>     this kind of dangerous misuse.

Reviewer 2: Not done?

---
minutes: 10
---

# Single-use values

In some circumstances we want values that _can only be used once_. One critical
example of this is in cryptography: "Nonces."

```rust,editable
pub struct Key(/* specifics omitted */);
// A single-use number suitable for cryptographic purposes.
pub struct Nonce(u32);
// A cryptographically sound random generator function.
pub fn new_nonce() -> Nonce {
    Nonce(4) // chosen by a fair dice roll, https://xkcd.com/221/
}

> // Consume a nonce, but not the key or the data.

Reviewer 2: ```suggestion
/// A single-use number suitable for cryptographic purposes.
pub struct Nonce(u32);
/// A cryptographically sound random generator function.
pub fn new_nonce() -> Nonce {
    Nonce(4) // chosen by a fair dice roll, https://xkcd.com/221/
}
/// Consume a nonce, but not the key or the data.
```

Three slashes for doc comments.

pub fn encrypt(nonce: Nonce, key: &Key, data: &[u8]) {}

fn main() {
    let nonce = new_nonce();
    let data_1: [u8; 4] = [1, 2, 3, 4];
    let data_2: [u8; 4] = [4, 3, 2, 1];
    let key = Key(/* specifics omitted */);

    // The key and data can be re-used, copied, etc. but the nonce cannot.
    encrypt(nonce, &key, &data_1);
    // encrypt(nonce, &key, &data_2); // 🛠️❌
}
```

<details>

> - What if we want to be able to guarantee that a value can only be used once?

Reviewer 2: ```suggestion
- Problem: How can we guarantee a value is used only once?
```

More concise rewrite, same content.
